File: tsd/tsd_cpp/core/types/instrument.cpp
#include "instrument.hpp"
#include <stdexcept>

namespace tsd {
namespace core {

std::string to_string(Currency currency) {
    switch (currency) {
        case Currency::BTC: return "BTC";
        case Currency::ETH: return "ETH";
        case Currency::USDT: return "USDT";
        case Currency::USDC: return "USDC";
        case Currency::USD: return "USD";
        case Currency::EUR: return "EUR";
        case Currency::GBP: return "GBP";
        case Currency::JPY: return "JPY";
        case Currency::BNB: return "BNB";
        case Currency::ADA: return "ADA";
        case Currency::DOT: return "DOT";
        case Currency::LINK: return "LINK";
        case Currency::SOL: return "SOL";
        case Currency::MATIC: return "MATIC";
        case Currency::AVAX: return "AVAX";
        case Currency::UNKNOWN: return "UNKNOWN";
        default: return "UNKNOWN";
    }
}

Currency currency_from_string(const std::string& str) {
    if (str == "BTC") return Currency::BTC;
    if (str == "ETH") return Currency::ETH;
    if (str == "USDT") return Currency::USDT;
    if (str == "USDC") return Currency::USDC;
    if (str == "USD") return Currency::USD;
    if (str == "EUR") return Currency::EUR;
    if (str == "GBP") return Currency::GBP;
    if (str == "JPY") return Currency::JPY;
    if (str == "BNB") return Currency::BNB;
    if (str == "ADA") return Currency::ADA;
    if (str == "DOT") return Currency::DOT;
    if (str == "LINK") return Currency::LINK;
    if (str == "SOL") return Currency::SOL;
    if (str == "MATIC") return Currency::MATIC;
    if (str == "AVAX") return Currency::AVAX;
    return Currency::UNKNOWN;
}

std::ostream& operator<<(std::ostream& os, const Currency& currency) {
    return os << to_string(currency);
}

std::ostream& operator<<(std::ostream& os, const Spot& spot) {
    return os << "Spot(" << spot.base << "/" << spot.term << ")";
}

std::ostream& operator<<(std::ostream& os, const Future& future) {
    return os << "Future(" << future.base << "/" << future.term << " " << future.expiry << ")";
}

std::ostream& operator<<(std::ostream& os, const Instrument& instrument) {
    switch (instrument.type) {
        case InstrumentType::SPOT:
            return os << instrument.spot;
        case InstrumentType::FUTURE:
            return os << instrument.future;
        default:
            return os << "Unknown Instrument";
    }
}

} // namespace core
} // namespace tsd

File: tsd/core/types/instrument.py
from datetime import datetime
from enum import StrEnum

import csp

from core.types.exchange import Exchange


class Instrument(csp.Struct):
    symbol: str


class Currency(StrEnum):
    USD = "USD"
    USDT = "USDT"
    USDC = "USDC"
    BTC = "BTC"
    ETH = "ETH"
    DAI = "DAI"
    SOL = "SOL"
    FDUSD = "FDUSD"
    XRP = "XRP"
    ZKJ = "ZKJ"
    DOGE = "DOGE"
    PEPE = "PEPE"
    SUI = "SUI"
    TRUMP = "TRUMP"


## Spot

class Spot(Instrument):
    base: Currency
    term: Currency

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.symbol = f"{self.base}{self.term}"

    def __str__(self):
        return self.symbol

    def __repr__(self):
        return self.symbol


## Future

# TODO: complete Enums and Structs for futures
# https://developers.binance.com/docs/derivatives/coin-margined-futures/common-definition
# https://developers.binance.com/docs/derivatives/usds-margined-futures/common-definition

class SymbolType(StrEnum):
    DELIVERY = "DELIVERY"
    PERPETUAL = "PERPETUAL"


class ContractType(StrEnum):
    PERPETUAL = "PERPETUAL"


class Future(Instrument):
    exchange: Exchange
    underlying: Instrument
    contract_size: float
    contract_price_multiplier: float
    min_price_increment: float
    is_inverted: bool
    is_primary: bool

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.name = f""


## Option

# TODO: complete Enums and Structs
# https://developers.binance.com/docs/derivatives/option/common-definition

class OptionStyle(StrEnum):
    EUROPEAN = "EUROPEAN"
    AMERICAN = "AMERICAN"


class OptionType(StrEnum):
    CALL = "CALL"
    PUT = "PUT"


class SettlementType(StrEnum):
    PHYSICAL = "PHYSICAL"
    CASH = "CASH"


class Option(Instrument):
    underlying: Instrument
    option_style: OptionStyle
    option_type: OptionType
    settlement_type: SettlementType
    settlement_currency: Currency
    strike: float
    expiration_date: datetime
    settlement_date: datetime

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.name = f""


if __name__ == "__main__":
    # instr = Instrument(symbol="DummyInstrument")
    # print(instr)
    #
    # btc_usd = Spot(base=Currency.BTC, term=Currency.USD)
    # print(btc_usd.term)
    # print(btc_usd.symbol)
    # print(btc_usd)
    print("BTC" in Currency)
    print(Currency.BTC in Currency)
